File: src/Program/ResolveModlist.cs
using System;
using System.IO;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using System.Linq;

namespace mcmli
{
    partial class Program
    {
        public static void ResolveModlist(string list)
        {
            // Download files in modlist
            string dir = "mods"; // If no directory given, download to "mods".
            bool alwaysFetch = false; // Skip existing files by default
            bool? clientMod = null; // All files are common by default
            List<string> mod_dirs = new List<String>();
            List<string> mods = new List<string>();

            foreach (string inputLine in File.ReadLines(list))
            {
                // Trim leading or trailing whitespace
                string line = inputLine.Trim();
                // Skip lines containing only whitespace
                if (String.IsNullOrWhiteSpace(line)) continue;

                bool isUrl = line.StartsWith("http");
                if (!isUrl)
                {
                    line = Regex.Replace(line, @"#.*", @""); // ignore comments

                    // Skip this line if is is not a valid control directive. That is,
                    // it doesn't contain anything inside square or angle brackets.
                    if (!Regex.Match(line, @"(\[[^]]+\]|<[^>]+>)").Success) continue;

                    if (ExtractFromDelims(line, "[]").Count > 0)
                    {
                        // Do this for every directory specification.
                        alwaysFetch = false;
                        clientMod = null;
                    }

                    var delimContent = ExtractFromDelims(line, "[]");
                    if (delimContent.Count > 0 && delimContent[0].ToString().Length != 0)
                        dir = delimContent[0].ToString();

                    if (ExtractFromDelims(line, "<>").Any(x =>
                          x.ToString().ToLower() == "always-fetch")) alwaysFetch = true;
                    if (ExtractFromDelims(line, "<>").Any(x =>
                          x.ToString().ToLower() == "smart-fetch")) alwaysFetch = false;

                    if (ExtractFromDelims(line, "<>").Any(x =>
                          x.ToString().ToLower() == "server-only")) clientMod = false;
                    if (ExtractFromDelims(line, "<>").Any(x =>
                          x.ToString().ToLower() == "client-only")) clientMod = true;
                    if (ExtractFromDelims(line, "<>").Any(x =>
                          x.ToString().ToLower() == "common")) clientMod = null;

                    continue;
                }

                // Add the current directory to the mod_dirs list if it's not
                // there already.
                if (!mod_dirs.Contains(dir)) mod_dirs.Add(dir);
                Directory.CreateDirectory(dir); // Make sure the directory exists

                ResolveUrl(line, dir, alwaysFetch, clientMod);
            }

            // Keep track of directories that we're storing mods in.
            foreach (string mod_dir in mod_dirs)
            {
                mods.AddRange(Directory.GetFiles(mod_dir, "*", SearchOption.TopDirectoryOnly).ToList());
            }

            foreach (string i in mods) // Remove local mods not in a modlist.
            {
                string filename;
                string local_filename = Path.GetFileName(i);
                string local_dir = Path.GetDirectoryName(i);

                // We want to re-encode filenames
                // NOTE: This may need to change if some mod urls contain unconventional
                //       characters.
                filename = local_filename.Replace(@" ", @"%20");
                filename = filename.Replace(@"+", @"%2B");
                filename = filename.Replace(@"#", @"%23");
                filename = filename.Replace(@":", @"%3A");
                filename = filename.Replace(@";", @"%3B");

                // Skip configs, old ones are harmless.
                if (filename.EndsWith("cfg", true, null)) continue;

                if (!modlists.Any(x => x != null &&
                      File.ReadAllText(x).Contains(filename)))
                {
                    Console.WriteLine("Found {0} in {1} directory but not in any modlist. It may be old.", local_filename, local_dir);

                    string resp;
                    string[] validResps = { "y", "n", "" };

                    do
                    {
                        Console.Write("Do you want to remove it? [Y/n]: ");
                        resp = Console.ReadLine().ToLower();
                    }
                    while (!validResps.Any(s => s.Equals(resp)));

                    if (resp == "y" || resp == "")
                    {
                        Console.WriteLine($"Removing {local_filename}");
                        File.Delete(i);
                    }
                }
            }
        }
    }
}

File: src/ResolveModlist.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"net/url"
	"os"
	"path"
	"path/filepath"
	"regexp"
	"strings"
)

// ResolveModlist comment
func ResolveModlist(Modlist string) {
	ResolveModlistAlways(Modlist, false)
}

// ResolveModlistAlways Resolve a modlist, optionally downloading everything
// regardless of whether it exists already.
func ResolveModlistAlways(Modlist string, always bool) {

	// Download to the current directory by default
	var DestDir string = "."
	// Reset alwaysFetch to this value for every directory specification
	var alwaysFetchDefault bool = always
	alwaysFetch = alwaysFetchDefault
	// Is this a common mod (-1), server only (0) or client only (1)
	var clientMod int = -1
	// Does the user need to download something manually?
	var userDownload bool = false

	// Regex that defines comments
	commentRegex := regexp.MustCompile(`#.*`)
	// Regex that defines comments in URLs
	commentRegexURL := regexp.MustCompile(` #.*`)
	// Regex that identifies a control seqeunce in the modlist.
	CSRegex := regexp.MustCompile(`(\[[^]]+\]|<[^>]+>)`)

	file, err := os.Open(Modlist)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		var Line string = strings.TrimSpace(scanner.Text())
		var isURL bool = strings.HasPrefix(Line, "http")
		if len(Line) == 0 { // Skip empty lines
			continue
		}

		if !isURL {
			Line = commentRegex.ReplaceAllString(Line, "") // ignore comments
			isCS := CSRegex.MatchString(Line)

			if !isCS { // If this line doesn't look like a control sequence, skip it
				continue
			}

			isDirectoryCS := len(ExtractFromDelims(Line, "[]")) > 0
			if isDirectoryCS { // Do this for every directory CS
				DestDir = ExtractFromDelims(Line, "[]")[0]
				// Reset alwaysFetch to its default value
				alwaysFetch = alwaysFetchDefault
				// Set mods to be common to both server and client by default
				clientMod = -1

				updating = false

				if !hasUpdated {
					// We'll definitely have updated by the time we get to a directory CS.
					hasUpdated = true
					// Now we wan't to recall this function with hasUpdated = true
					ResolveModlistAlways(Modlist, always)
					// Reset hasUpdated in case there are other modlists that need to update
					// themselves.
					hasUpdated = false
					// We're all done with this modlist after the recursion above, so just return
					return
				}
			}

			// Check for client/server only and alwaysFetch
			for _, cs := range ExtractFromDelims(Line, "<>") {
				if strings.ToLower(cs) == "server-only" {
					clientMod = 0
				}
				if strings.ToLower(cs) == "client-only" {
					clientMod = 1
				}
				if strings.ToLower(cs) == "common" {
					clientMod = -1
				}
				if strings.ToLower(cs) == "always-fetch" {
					alwaysFetch = true
				}
				if strings.ToLower(cs) == "smart-fetch" {
					alwaysFetch = false
				}
				if strings.ToLower(cs) == "update" {
					/* URLs should always be downloaded when updating. They should also be
					updated for both client and server */
					alwaysFetch = true
					clientMod = -1
					updating = true
				}
			}
			// That's all we need to do for anything that's not a URL, so just move on
			// to the next line.
			continue
		}

		Line = commentRegexURL.ReplaceAllString(Line, "")

		Filename, err := url.PathUnescape(path.Base(Line))
		if err != nil {
			log.Fatal(err)
		}

		// If this file is client only and the current instance is a server
		if clientMod == 1 && isServer {
			// Remove the file if it exists
			if PathExists(filepath.Join(DestDir, Filename)) {
				fmt.Printf("[%s] Removing client only file: %s\n", DestDir, Filename)
				os.Remove(filepath.Join(DestDir, Filename))
			}
			// Go to the next line in LinesInModlist
			continue
		}
		// As above
		if clientMod == 0 && !isServer {
			if PathExists(filepath.Join(DestDir, Filename)) {
				fmt.Printf("[%s] Removing server only file: %s\n", DestDir, Filename)
				os.Remove(filepath.Join(DestDir, Filename))
			}
			continue
		}

		// Create the destination directory.
		os.MkdirAll(DestDir, 0755)

		if updating {
			if hasUpdated {
				// Go to the next line if we've already updated
				continue
			} else {
				fmt.Println("Updating ...")
				DownloadFile(Line, DestDir)
				continue
			}
		}

		if !ContainsStr(modDirs, DestDir) && DestDir != "." && DestDir != "config" {
			modDirs = append(modDirs, DestDir)
		}

		if useCache && !always && !alwaysFetch {
			GetFileFromCache(Line, DestDir)
			continue
		}

		if alwaysFetch || !PathExists(filepath.Join(DestDir, Filename)) {
			DownloadFile(Line, DestDir)
		} else {
			fmt.Printf("[%s] Skipping existing file %s\n", DestDir, Filename)
		}

	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	// Slice containing all mods
	var mods []string
	for _, i := range modDirs {
		modfiles, _ := GetFilesInDirectory(i)
		for j, k := range modfiles {
			modfiles[j] = filepath.Join(i, k)
		}
		// Append this list of mods to the main slice.
		mods = append(mods, modfiles...)
	}

	for _, mod := range mods {
		var isInModlist bool = false
		// The directory that the mod is in
		dir := filepath.Dir(mod)
		encodedMod := url.PathEscape(filepath.Base(mod))
		// We want to decode some strings because of inconsistencies.
		encodedMod = strings.Replace(encodedMod, "%27", "'", -1)

		// For each modlist, check if any line in it contains the escaped mod filename
		for _, modlist := range modlists {
			modlistLines, _ := ReadLines(modlist)
			for _, line := range modlistLines {
				if strings.HasSuffix(line, encodedMod) {
					isInModlist = true
					// No point doing any further checking for this mod
					goto NEXTMOD
				}
			}
		}
		// If a file is found locally, but it's not in a modlist: prompt the user to
		// remove it.
		if !isInModlist {
			fmt.Printf("%s found in '%s' directory but not in any modlist.\n", encodedMod, dir)
			resp := GetUserInput("Do you want to remove it? [Y/n]: ")
			if strings.ToLower(resp) == "y" || resp == "" {
				os.Remove(mod)
			}
		}
	NEXTMOD:
	}
}
